lib/rucio/core/did_meta_plugins/__init__.py

```diff
@@ -16,157 +16,194 @@
+# Set default modules.
+#
+# - "base" metadata is tied to column fields in the "dids" table. As such, types are restricted to column types and
+#   additional fields cannot be created.
+# - "custom" metadata refers to any plugin that has been written to interface with Rucio's metadata plugin system.
+#
+# By default, the default custom metadata plugin is set to the prepackaged json module. This utilises a separate table,
+# "did_meta", allowing users to add custom metadata key/value pairs of any type.
+#
+DEFAULT_BASE_METADATA_PLUGIN_MODULE_PATH = "rucio.core.did_meta_plugins.did_column_meta.DidColumnMeta"
+DEFAULT_CUSTOM_METADATA_PLUGIN_MODULE_PATH = "rucio.core.did_meta_plugins.json_meta.JSONDidMeta"
 
+# Overwrite these defaults if plugins are set in the configuration file.
+#
 if config.config_has_section('metadata'):
     try:
-        METADATA_HANDLER_MODULES = config.config_get('metadata', 'plugins')
+        CUSTOM_METADATA_PLUGIN_MODULE_PATHS = config.config_get('metadata', 'plugins')
     except (NoOptionError, NoSectionError):
-        METADATA_HANDLER_MODULES = FALLBACK_METADATA_HANDLER_MODULE
+        CUSTOM_METADATA_PLUGIN_MODULE_PATHS = DEFAULT_CUSTOM_METADATA_PLUGIN_MODULE_PATH
 else:
-    METADATA_HANDLER_MODULES = FALLBACK_METADATA_HANDLER_MODULE
+    CUSTOM_METADATA_PLUGIN_MODULE_PATHS = DEFAULT_CUSTOM_METADATA_PLUGIN_MODULE_PATH
+METADATA_PLUGIN_MODULE_PATHS = [DEFAULT_BASE_METADATA_PLUGIN_MODULE_PATH] + CUSTOM_METADATA_PLUGIN_MODULE_PATHS.split(",")
 
-META_MODULE_PATHS = [HARDCODED_METADATA_HANDLER_MODULE] + METADATA_HANDLER_MODULES.split(",")
-
-METADATA_HANDLERS = []
-for meta_module_path in META_MODULE_PATHS:
+# Import plugin modules.
+#
+# Note that the order of this list is important. As the base metadata plugin module is always first, base key
+# retrieval and setting will always take precedence over custom plugins, i.e. it is not possible to set a custom key with
+# the same name as those in the base list.
+#
+# Another consequence of this is that if set_metadata() is called with multiple plugins specified, the first to return
+# True to manages_key() will be used.
+#
+METADATA_PLUGIN_MODULES = []
+for meta_module_path in METADATA_PLUGIN_MODULE_PATHS:
     try:
         base_module = ".".join(meta_module_path.split(".")[:-1])
         base_class = meta_module_path.split(".")[-1]
-        meta_handler_module = getattr(importlib.import_module(base_module), base_class)()
-        METADATA_HANDLERS.append(meta_handler_module)
+        metadata_plugin_module = getattr(importlib.import_module(base_module), base_class)()
+        METADATA_PLUGIN_MODULES.append(metadata_plugin_module)
     except ImportError:
-        raise PolicyPackageNotFound('Module ' + meta_module_path + ' not found')
+        raise exception.PolicyPackageNotFound('Module ' + meta_module_path + ' not found')
+
+# Set restricted character set for metadata in form character: reason
+#
+RESTRICTED_CHARACTERS = {
```

Comment:
Disallow certain characters in metadata. '.' is used as a delimiter between the key and operator.